REMEMBER: FIRST CREATE A COPY OF THIS FILE WITH A UNIQUE NAME AND DO YOUR WORK THERE. AND MAKE SURE YOU COMMIT YOUR CHANGES TO THE `hw3_submissions` BRANCH.

# Assignment 3 | Cleaning and Exploring Data with Pandas



<img src="data/scoreCard.jpg" width=250>

In this assignment, you will investigate restaurant food safety scores for restaurants in San Francisco. Above is a sample score card for a restaurant. The scores and violation information have been made available by the San Francisco Department of Public Health. 

## Loading Food Safety Data


There are 2 files in the data directory:
1. business.csv containing food establishments in San Francisco
1. inspections.csv containing retaurant inspections records

Let's start by loading them into Pandas dataframes.  One of the files, business.csv, has encoding (ISO-8859-1), so you will need to account for that when reading it.

### Question 1

#### Question 1a
Read the two files noted above into two pandas dataframes named `bus` and `ins`, respectively. Print the first 5 rows of each to inspect them.


In [1]:
import pandas as pd

In [2]:
bus = pd.read_csv('data/businesses.csv', encoding='ISO-8859-1')
ins = pd.read_csv('data/inspections.csv')

In [3]:
bus.head()

business_id                                name  \
0           10                    TIRAMISU KITCHEN   
1           19               NRGIZE LIFESTYLE CAFE   
2           24  OMNI S.F. HOTEL - 2ND FLOOR PANTRY   
3           31      NORMAN'S ICE CREAM AND FREEZES   
4           45                 CHARLIE'S DELI CAFE   

                         address           city state postal_code   latitude  \
0                  033 BELDEN PL  San Francisco    CA       94104  37.791116   
1   1200 VAN NESS AVE, 3RD FLOOR  San Francisco    CA       94109  37.786848   
2  500 CALIFORNIA ST, 2ND  FLOOR  San Francisco    CA       94104  37.792888   
3           2801 LEAVENWORTH ST   San Francisco    CA       94133  37.807155   
4                3202 FOLSOM ST   San Francisco    CA       94110  37.747114   

    longitude  phone_number  
0 -122.403816  +14154217044  
1 -122.421547  +14157763262  
2 -122.403135  +14156779494  
3 -122.419004           NaN  
4 -122.413641  +14156415051

In [4]:
ins.head()

business_id  score      date     type
0           10     82  20160503  routine
1           10     94  20140729  routine
2           10     92  20140114  routine
3           19     94  20160513  routine
4           19     94  20141110  routine

## Examining the Business data

From its name alone, we expect the `businesses.csv` file to contain information about the restaurants. Let's investigate this dataset.

### Question 2

#### Question 2a: How many records are there?

In [16]:
bus['business_id'].count()

6315

In [17]:
ins['business_id'].count()

15430

#### Question 2b: How many unique business IDs are there?  

In [9]:
len(bus['business_id'].unique())

6315

In [15]:
len(ins['business_id'].unique())

5730

#### Question 2c: What are the 5 most common businesses by name, and how many are there in San Francisco?

In [12]:
bus['name'].value_counts()

STARBUCKS COFFEE              72
PEET'S COFFEE & TEA           24
MCDONALDS                     12
SAN FRANCISCO SOUP COMPANY    11
WALGREENS                     11
                              ..
7-MISSION RESTAURANT           1
WONDERFUL DESSERT & CAFE       1
KITCHEN ISTANBAL               1
FAT BELLI DELI                 1
ALBA RAY'S                     1
Name: name, Length: 5692, dtype: int64

## Zip code

Next, let's  explore some of the variables in the business table. We begin by examining the postal code.

### Question 3

#### Question 3a
How are the zip code values stored in python (i.e. data type)?

To answer this you might want to examine a particular entry.

In [22]:
bus['postal_code'].describe()

count      6056
unique       46
top       94110
freq        585
Name: postal_code, dtype: object

#### Question 3b

What are the unique values of postal_code?

In [24]:
bus['postal_code'].unique()

array(['94104', '94109', '94133', '94110', '94122', '94103', '94115',
       '94131', '94111', '94117', '94107', '94108', '94102', '94132',
       '94105', '94134', '94124', '94116', '94121', '94118', '94112',
       '94123', '94114', '94127', '941102019', '941', '94545', '94066',
       '941033148', nan, '94158', '95105', '94013', '94130', 'CA',
       '92672', '94120', '94143', '94609', '94101', '00000', '94188',
       '94621', '94014', '94129', '94602', 'Ca'], dtype=object)

#### Question 3c

Let's say we decide to exclude the businesses that have no zipcode for our analysis (which might include food trucks for example).  Use the list of valid 5-digit zip codes below to create a new dataframe called bus_valid, with only businesses whose postal_codes show up in this list of valid zipcodes. How many businesses are there in this new dataframe?

In [25]:
validZip = ["94102", "94103", "94104", "94105", "94107", "94108",
            "94109", "94110", "94111", "94112", "94114", "94115",
            "94116", "94117", "94118", "94121", "94122", "94123", 
            "94124", "94127", "94131", "94132", "94133", "94134"]

In [26]:
vb = bus[bus['postal_code'].isin(validZip)]

In [28]:
vb.head()

business_id                                name  \
0           10                    TIRAMISU KITCHEN   
1           19               NRGIZE LIFESTYLE CAFE   
2           24  OMNI S.F. HOTEL - 2ND FLOOR PANTRY   
3           31      NORMAN'S ICE CREAM AND FREEZES   
4           45                 CHARLIE'S DELI CAFE   

                         address           city state postal_code   latitude  \
0                  033 BELDEN PL  San Francisco    CA       94104  37.791116   
1   1200 VAN NESS AVE, 3RD FLOOR  San Francisco    CA       94109  37.786848   
2  500 CALIFORNIA ST, 2ND  FLOOR  San Francisco    CA       94104  37.792888   
3           2801 LEAVENWORTH ST   San Francisco    CA       94133  37.807155   
4                3202 FOLSOM ST   San Francisco    CA       94110  37.747114   

    longitude  phone_number  
0 -122.403816  +14154217044  
1 -122.421547  +14157763262  
2 -122.403135  +14156779494  
3 -122.419004           NaN  
4 -122.413641  +14156415051

In [30]:
len(vb['business_id'].unique())

5999

## Latitude and Longitude

Another aspect of the data we want to consider is the prevalence of missing values. If many records have missing values then we might be concerned about whether the nonmissing values are representative of the population.

### Question 4
 
Consider the longitude and latitude in the business DataFrame. 

#### Question 4a

How many businesses are missing longitude values, working with only the businesses that are in the list of valid zipcodes?

In [35]:
vb[pd.isnull(vb['longitude'])]

business_id                               name             address  \
200           821                       LA SANTANECA    3781 MISSION ST    
210           852  NEW HOLLYWOOD BAKERY & RESTAURANT    652 PACIFIC AVE    
760          2934                   LOU'S FISH SHACK   300 JEFFERSON ST    
826          3123                        LES JOULINS        44 ELLIS ST    
1100         4323                             DIVA'S       1081 POST ST    
...           ...                                ...                 ...   
6310        90234               ACCESS SFUSD THE ARC     1520 HOWARD ST    
6311        90236                     PHO DONG HUONG  667 MONTEREY BLVD    
6312        90237                   WIN'S RESTAURANT    3040 TARAVAL ST    
6313        90238                   USA POWER MARKET     1524 OCEAN AVE    
6314        90239                         ALBA RAY'S    2293 MISSION ST    

               city state postal_code  latitude  longitude  phone_number  
200   San Francisco    CA       94110       NaN        NaN  +14156481034  
210   San Francisco    CA       94133       NaN        NaN  +14153979919  
760   San Francisco    CA       94133       NaN        NaN  +14157715687  
826   San Francisco    CA       94102       NaN        NaN  +14153975397  
1100  San Francisco    CA       94109       NaN        NaN  +14154743482  
...             ...   ...         ...       ...        ...           ...  
6310  San Francisco    CA       94103       NaN        NaN           NaN  
6311  San Francisco    CA       94127       NaN        NaN  +14154945048  
6312  San Francisco    CA       94116       NaN        NaN  +14158606036  
6313  San Francisco    CA       94112       NaN        NaN  +14153331878  
6314  San Francisco    CA       94110       NaN        NaN  +14158881393  

[2483 rows x 9 columns]

#### Question 4b

Create a new dataframe with one row for each valid zipcode. The dataframe should include the following three columns:

1. `postal_code`: Contains the zip codes in the `validZip` variable above.
2. `null_lon`: The number of businesses in that zipcode with missing `longitude` values.
3. `not_null_lon`: The number of businesses without missing `longitude` values.

In [109]:
vb_null = vb[pd.isnull(vb['longitude'])]

In [110]:
vb_null_grouped = vb_null[['business_id', 'postal_code']].groupby('postal_code')
vb_null_grouped.count()

business_id
postal_code             
94102                196
94103                252
94104                 59
94105                100
94107                247
94108                 82
94109                146
94110                251
94111                115
94112                 59
94114                108
94115                 76
94116                 35
94117                 67
94118                100
94121                 55
94122                123
94123                 53
94124                 88
94127                 28
94131                 14
94132                 65
94133                135
94134                 29

In [111]:
vb_final = vb_null_grouped.count()

In [112]:
vb_final.head()

business_id
postal_code             
94102                196
94103                252
94104                 59
94105                100
94107                247

In [113]:
vb_notnull = vb[pd.notnull(vb['longitude'])]

In [114]:
vb_notnull_grouped = vb_notnull[['business_id', 'postal_code']].groupby('postal_code')
vb_notnull_grouped.count()

business_id
postal_code             
94102                262
94103                306
94104                 74
94105                126
94107                204
94108                142
94109                237
94110                334
94111                167
94112                131
94114                122
94115                162
94116                 58
94117                130
94118                143
94121                106
94122                147
94123                116
94124                 76
94127                 43
94131                 34
94132                 69
94133                282
94134                 45

In [115]:
vb_final['longitude_notnull'] = vb_notnull_grouped.count()
vb_final.head()

business_id  longitude_notnull
postal_code                                
94102                196                262
94103                252                306
94104                 59                 74
94105                100                126
94107                247                204

In [119]:
vb_final = vb_final.rename(columns={'business_id': 'longitude_null'})
vb_final.head(30)

longitude_null  longitude_notnull
postal_code                                   
94102                   196                262
94103                   252                306
94104                    59                 74
94105                   100                126
94107                   247                204
94108                    82                142
94109                   146                237
94110                   251                334
94111                   115                167
94112                    59                131
94114                   108                122
94115                    76                162
94116                    35                 58
94117                    67                130
94118                   100                143
94121                    55                106
94122                   123                147
94123                    53                116
94124                    88                 76
94127                    28                 43
94131                    14                 34
94132                    65                 69
94133                   135                282
94134                    29                 45

#### 4c. Do any zip codes appear to have more than their 'fair share' of missing longitude? 

To answer this, you will want to compute the proportion of missing longitude values for each zip code, and print the proportion missing longitude, and print the top five zipcodes in descending order of proportion missing postal_code.


In [121]:
vb_final['prop'] = (vb_final['longitude_null']/(vb_final['longitude_null']+vb_final['longitude_notnull']))
vb_final.head(30)

longitude_null  longitude_notnull      prop
postal_code                                             
94102                   196                262  0.427948
94103                   252                306  0.451613
94104                    59                 74  0.443609
94105                   100                126  0.442478
94107                   247                204  0.547672
94108                    82                142  0.366071
94109                   146                237  0.381201
94110                   251                334  0.429060
94111                   115                167  0.407801
94112                    59                131  0.310526
94114                   108                122  0.469565
94115                    76                162  0.319328
94116                    35                 58  0.376344
94117                    67                130  0.340102
94118                   100                143  0.411523
94121                    55                106  0.341615
94122                   123                147  0.455556
94123                    53                116  0.313609
94124                    88                 76  0.536585
94127                    28                 43  0.394366
94131                    14                 34  0.291667
94132                    65                 69  0.485075
94133                   135                282  0.323741
94134                    29                 45  0.391892

In [123]:
vb_final.sort_values(['prop'], ascending=False).head(5)

longitude_null  longitude_notnull      prop
postal_code                                             
94107                   247                204  0.547672
94124                    88                 76  0.536585
94132                    65                 69  0.485075
94114                   108                122  0.469565
94122                   123                147  0.455556

# Investigate the inspection data

Let's now turn to the inspection DataFrame. Earlier, we found that `ins` has 4 columns, these are named `business_id`, `score`, `date` and `type`.  In this section, we determine the granularity of `ins` and investigate the kinds of information provided for the inspections. 

### Question 5

#### Question 5a
As with the business data, assess whether there is one inspection record for each business, by counting how many rows are in the data and how many unique businesses there are in the data. If they are exactly the same number, it means there is only one inspection per business, clearly.

In [50]:
ins['business_id'].count()

15430

In [51]:
len(ins['business_id'].unique())

5730

#### Question 5b

What values does `type` take on? How many occurrences of each value is in the DataFrame? Create a new dataframe named `ins2` by copying `ins` and keeping only records with values of `type` that occur more than 10 times in the original table. In other words, eliminate records that have values of `type` that occur rarely (< 10 times). Check the result to make sure rare types are eliminated.

In [53]:
ins['type'].unique()

array(['routine', 'complaint'], dtype=object)

In [54]:
ins['type'].value_counts()

routine      15429
complaint        1
Name: type, dtype: int64

In [56]:
ins2 = ins[(ins['type'] == 'routine')]

In [57]:
ins2['type'].value_counts()

routine    15429
Name: type, dtype: int64

#### Question 5c

Since the data was stored in a .csv file, the dates are formatted as strings such as `20160503`. Once we read in the data, we would like to have dates in an appropriate format for analysis. Add a new column called `year` by capturing the first four characters of the date column. 

Hint: we have seen multiple ways of doing this in class, includings `str` operations, `lambda` functions, `datetime` operations, and others. Choose the method that works best for you :)

In [62]:
ins2['date'].describe()

count    1.542900e+04
mean     2.015071e+07
std      8.720871e+03
min      2.013123e+07
25%      2.014072e+07
50%      2.015060e+07
75%      2.016052e+07
max      2.016122e+07
Name: date, dtype: float64

In [64]:
ins2['year'] = ins2['date'].astype(str).str[:4]

/var/folders/7p/z4qnh93942q52s0bs8hkz6yw0000gn/T/ipykernel_54193/3786169749.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  ins2['year'] = ins2['date'].astype(str).str[:4]


In [65]:
ins2.head()

business_id  score      date     type  year
0           10     82  20160503  routine  2016
1           10     94  20140729  routine  2014
2           10     92  20140114  routine  2014
3           19     94  20160513  routine  2016
4           19     94  20141110  routine  2014

#### Question 5d

What range of years is covered in this data set? Are there roughly same number of inspections each year? Try dropping records for any years with less than 50 inspections and store the result in a new dataframe named `ins3`.

In [67]:
ins2['year'].value_counts()

2016    5839
2014    5629
2015    3923
2013      38
Name: year, dtype: int64

In [71]:
ins3 = ins2[(ins2['year'] != '2013')]

In [72]:
ins3['year'].value_counts()

2016    5839
2014    5629
2015    3923
Name: year, dtype: int64

Let's examine only the inspections for one year: 2016. This puts businesses on a more equal footing because [inspection guidelines](https://www.sfdph.org/dph/eh/Food/Inspections.asp) generally refer to how many inspections should occur in a given year.

In [73]:
ins3[(ins3['year'] == '2016')]

business_id  score      date     type  year
0               10     82  20160503  routine  2016
3               19     94  20160513  routine  2016
6               24     98  20161005  routine  2016
7               24     96  20160311  routine  2016
10              31     84  20160816  routine  2016
...            ...    ...       ...      ...   ...
15425        89515     98  20161103  routine  2016
15426        89701    100  20161206  routine  2016
15427        89790     96  20161129  routine  2016
15428        89803    100  20161215  routine  2016
15429        89900     93  20161206  routine  2016

[5839 rows x 5 columns]

### Question 6

#### Question 6a

Merge the business and 2016 inspections data, keeping all businesses regardless of whether they show up in the inspections file. Show the first several rows of the resulting dataframe.

In [90]:
bus_ins = pd.merge(bus, ins3, left_on='business_id', right_on='business_id', how='left')

In [91]:
bus_ins.head(20)

business_id                                name  \
0            10                    TIRAMISU KITCHEN   
1            10                    TIRAMISU KITCHEN   
2            10                    TIRAMISU KITCHEN   
3            19               NRGIZE LIFESTYLE CAFE   
4            19               NRGIZE LIFESTYLE CAFE   
5            19               NRGIZE LIFESTYLE CAFE   
6            24  OMNI S.F. HOTEL - 2ND FLOOR PANTRY   
7            24  OMNI S.F. HOTEL - 2ND FLOOR PANTRY   
8            24  OMNI S.F. HOTEL - 2ND FLOOR PANTRY   
9            24  OMNI S.F. HOTEL - 2ND FLOOR PANTRY   
10           31      NORMAN'S ICE CREAM AND FREEZES   
11           31      NORMAN'S ICE CREAM AND FREEZES   
12           31      NORMAN'S ICE CREAM AND FREEZES   
13           45                 CHARLIE'S DELI CAFE   
14           45                 CHARLIE'S DELI CAFE   
15           45                 CHARLIE'S DELI CAFE   
16           48                          ART'S CAFE   
17           48                          ART'S CAFE   
18           50                          SUSHI ZONE   
19           50                          SUSHI ZONE   

                          address           city state postal_code   latitude  \
0                   033 BELDEN PL  San Francisco    CA       94104  37.791116   
1                   033 BELDEN PL  San Francisco    CA       94104  37.791116   
2                   033 BELDEN PL  San Francisco    CA       94104  37.791116   
3    1200 VAN NESS AVE, 3RD FLOOR  San Francisco    CA       94109  37.786848   
4    1200 VAN NESS AVE, 3RD FLOOR  San Francisco    CA       94109  37.786848   
5    1200 VAN NESS AVE, 3RD FLOOR  San Francisco    CA       94109  37.786848   
6   500 CALIFORNIA ST, 2ND  FLOOR  San Francisco    CA       94104  37.792888   
7   500 CALIFORNIA ST, 2ND  FLOOR  San Francisco    CA       94104  37.792888   
8   500 CALIFORNIA ST, 2ND  FLOOR  San Francisco    CA       94104  37.792888   
9   500 CALIFORNIA ST, 2ND  FLOOR  San Francisco    CA       94104  37.792888   
10           2801 LEAVENWORTH ST   San Francisco    CA       94133  37.807155   
11           2801 LEAVENWORTH ST   San Francisco    CA       94133  37.807155   
12           2801 LEAVENWORTH ST   San Francisco    CA       94133  37.807155   
13                3202 FOLSOM ST   San Francisco    CA       94110  37.747114   
14                3202 FOLSOM ST   San Francisco    CA       94110  37.747114   
15                3202 FOLSOM ST   San Francisco    CA       94110  37.747114   
16                 747 IRVING ST   San Francisco    CA       94122  37.764013   
17                 747 IRVING ST   San Francisco    CA       94122  37.764013   
18              1815  MARKET ST.   San Francisco    CA       94103  37.771437   
19              1815  MARKET ST.   San Francisco    CA       94103  37.771437   

     longitude  phone_number  score        date     type  year  
0  -122.403816  +14154217044   82.0  20160503.0  routine  2016  
1  -122.403816  +14154217044   94.0  20140729.0  routine  2014  
2  -122.403816  +14154217044   92.0  20140114.0  routine  2014  
3  -122.421547  +14157763262   94.0  20160513.0  routine  2016  
4  -122.421547  +14157763262   94.0  20141110.0  routine  2014  
5  -122.421547  +14157763262   94.0  20140214.0  routine  2014  
6  -122.403135  +14156779494   98.0  20161005.0  routine  2016  
7  -122.403135  +14156779494   96.0  20160311.0  routine  2016  
8  -122.403135  +14156779494   96.0  20141124.0  routine  2014  
9  -122.403135  +14156779494   96.0  20140612.0  routine  2014  
10 -122.419004           NaN   84.0  20160816.0  routine  2016  
11 -122.419004           NaN   98.0  20151204.0  routine  2015  
12 -122.419004           NaN  100.0  20140908.0  routine  2014  
13 -122.413641  +14156415051   84.0  20160614.0  routine  2016  
14 -122.413641  +14156415051   78.0  20160104.0  routine  2016  
15 -122.413641  +14156415051   94.0  20140909.0  routine  2014  
16 -122.465749  +14156657440   94.0  2016

#### Question 6b
Print the 20 lowest rated businesses names, their addresses, and their ratings.

In [93]:
bus_ins_worst = bus_ins[['name', 'address', 'score']]

In [94]:
bus_ins_worst.head()

name                       address  score
0       TIRAMISU KITCHEN                 033 BELDEN PL   82.0
1       TIRAMISU KITCHEN                 033 BELDEN PL   94.0
2       TIRAMISU KITCHEN                 033 BELDEN PL   92.0
3  NRGIZE LIFESTYLE CAFE  1200 VAN NESS AVE, 3RD FLOOR   94.0
4  NRGIZE LIFESTYLE CAFE  1200 VAN NESS AVE, 3RD FLOOR   94.0

In [95]:
bus_ins_worst.sort_values(['score'], ascending=True).head(20)

name                     address  \
10664                        ABC BAKERY CAFE             650 JACKSON ST    
6362                         IMPERIAL PALACE          818 WASHINGTON ST    
9258          YUMMY DIM SUM & FAST FOOD, LLC            930 STOCKTON ST    
11708                GOLDEN RIVER RESTAURANT            5827 GEARY BLVD    
2446                             LES JOULINS                44 ELLIS ST    
2426                        YEE'S RESTAURANT             1131 GRANT AVE    
6710                     BRISTOL FARMS #2317           845 MARKET ST 010   
5907                           BODEGA BISTRO              607 LARKIN ST    
8629        OZONE THAI RESTAURANT AND LOUNGE                598 02ND ST    
2715             BASIL THAI RESTAURANT & BAR             1175 FOLSOM ST    
12076                        MISS SAIGON LLC                100 06TH ST    
11665                             UNCLE CAFE              65 WAVERLY PL    
14587                HARBOR VILLA RESTAURANT  5238 DIAMOND HEIGHTS BLVD    
7749                              FIRST CAKE            5901 GEARY BLVD    
15689                              POKI TIME            2101 LOMBARD ST    
14455                           CRAZY PEPPER          2257 SAN JOSE AVE    
11651                          SA BEANG THAI          312 DIVISADERO ST    
3362                         GOURMET KITCHEN           1051 STOCKTON ST    
14400  LE SOLEIL INTERNATIONAL HOLDINGS, INC          133 CLEMENT  ST.     
858                          TAQUERIA LA PAZ                305 TURK ST    

       score  
10664   46.0  
6362    47.0  
9258    50.0  
11708   52.0  
2446    53.0  
2426    53.0  
6710    54.0  
5907    54.0  
8629    54.0  
2715    54.0  
12076   54.0  
11665   55.0  
14587   55.0  
7749    55.0  
15689   55.0  
14455   55.0  
11651   55.0  
3362    55.0  
14400   55.0  
858     56.0

## Done!

Now commit this notebook to your `hw3_submissions` branch, push it to your GitHub repo, and open a PR!